pyClient/test/test_contract_base_mixer.py

```diff
@@ -0,0 +1,314 @@
+
+# The variable inputs represents a dummy primary input array,
+# it is structured as follows,
+# inputs =
+#   rt || {sn}_1,2 || {cm}_1,2 || h_sig || {h}_1,2 || residual_bits
+# residual_bits =
+#   v_in || v_out || h_sig || {sn}_1,2  || {cm}_1,2  || {h}_1,2
+# We set dummy values for all variables but residual_bits as follows:
+# residual_bits = 713623846352979940490457358497079434602616037, or in bits
+# 1-4:   00000000 00000000 00000000 00000000
+# 5-8:   00000000 00000000 00000000 00000000
+# 9-12:  00000000 00000000 00000000 00000000
+# 13-16: 00000000 00011111 11111111 11111111
+# 17-20: 11111111 11111111 11111111 11111111
+# 21-24: 11111111 11100000 00000000 00000000
+# 25-28: 00000000 00000000 00000000 00000000
+# 29-32: 00000000 00011100 00010100 11100101
+# This corresponds to
+# v_in  = 0xFFFFFFFFFFFFFFFF
+# v_out = 0x0000000000000000
+# h_sig = 111
+# sn_0  = 000
+# sn_1  = 001
+# cm_0  = 010
+# cm_1  = 011
+# h_0   = 100
+# h_1   = 101
+# The values were set to be easily distinguishable.
+inputs = [
+    0,  # root
+    1,  # sn_0
+    1,  # sn_1
+    2,  # cm_0
+    2,  # cm_1
+    3,  # h_sig
+    4,  # h_0
+    4,  # h_1
+    713623846352979940490457358497079434602616037] \
+        # residual bits ; pylint: disable=no-member,invalid-name
+
+
+def test_extract(mixer_instance: Any) -> int:
+    # This test checks that we extract the right values
+    # for {sn}, {cm} and {h} regardless of whether
+    # the bits are located in the same byte
+    # or two different ones (c.f. residual_bits above).
+    print("--- testing ", "test_extract")
+
+    start = 0
+    for i in range(2*js_in+js_out):
+        res = mixer_instance.functions.extract_extra_bits(
+            start,
+            packing_residue_length,
+            inputs
+            ).call()
+        if int.from_bytes(res, byteorder="big") != i:
+            print("ERROR: extracted wrong value")
+            return 0
+        start += packing_residue_length
+    return 1
+
+
+def test_assemble_nullifiers(mixer_instance: Any) -> int:
+    # Test retrieving nullifiers
+    print("--- testing ", "test_assemble_nullifiers")
+    for i in range(js_in):
+        res = mixer_instance.functions.assemble_nullifier(i, inputs).call()
+        val = int.from_bytes(res, byteorder="big")
+        expected_val = 2**(256-1) + int(("{0:03b}".format(i))[::-1], 2)
+        if val != expected_val:
+            print("ERROR: extracted wrong nullifier")
+            print("expected:", expected_val, i)
+            print("got:", val, i)
+            return 0
+    return 1
+
+
+def test_assemble_commitments(mixer_instance: Any) -> int:
+    # Test retrieving commitments
+    print("--- testing ", "test_assemble_commitments")
+    for i in range(js_out):
+        res = mixer_instance.functions.assemble_commitment(i, inputs).call()
+        val = int.from_bytes(res, byteorder="big")
+        expected_val = 2**(256-1-1) + int(("{0:03b}".format(2+i))[::-1], 2)
```

Comment:
why `2**(256-1-1)`, and not `2**(256-1)`? the expected value should be a digest of 256 bits, so the leading bit should correspond to `2**(256-1)`

Reply:
Hum... ok I think I understand now... why not just define the `expected_var` from the `inputs` you define at the beginning of the document instead of doing it from the loop index. It's just confusing.

Reply:
Updated with `inputs[ ]`